### PR TITLE
[Dynamo] Add UserError type

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97705

To get started the dynamo error message improvement effort, we discussed about adding new user error type which covers cases where the user used something that TorchDynamo doesn't support and there is clear actions they can take.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/exc.py [hunk 1]
```diff
@@ -73,6 +74,26 @@ def add_to_stats(self, category="unimplemented"):
...
+class UserErrorType(Enum):
+    DYNAMIC_CONTROL_FLOW = auto()
+
+
+# TODO Ideally it should inherit from UnsupportedError
+# because if user code doesn't compile, we know it is not
+# supported. But seperate exception for now as to
+# not pollute the stat collector.
+class UserError(TorchDynamoException):
+    def __init__(self, error_type: UserErrorType, msg):
```
Comment (anijain2305) [inline torch/_dynamo/exc.py:86]: Please add comment here explaining UserError (same one that you have in the description).

**Review by anijain2305 — APPROVED** [review]

**Review by yanboliang — APPROVED** [review]

File: torch/_dynamo/exc.py (hunk 1 above)
Comment (tugsbayasgalan) [reply torch/_dynamo/exc.py:86]: done

File: torch/_dynamo/symbolic_convert.py [hunk 2]
```diff
@@ -329,7 +329,12 @@ def inner(self: "InstructionTranslatorBase", inst: Instruction):
                 push and self.push(value)
                 self.jump(inst)
         else:
-            unimplemented(f"generic_jump {typestr(value)}")
+            # TODO link the torch.cond doc later
+            raise exc.UserError(
+                exc.UserErrorType.DYNAMIC_CONTROL_FLOW,
+                "Dynamic control flow is not supported at the moment. Please use "
```
Comment (voznesenskym) [inline torch/_dynamo/symbolic_convert.py:335]: It's not dynamic control flow, it's data dependent control flow.

**Pull request closed without merging**